### DISCUSSION TITLE
Is there a way to find from ROS2 driver if I am using HW or URSim

### DISCUSSION THREAD

**Discussion by jsubramo** [UniversalRobots/Universal_Robots_ROS2_Driver#744] [General]
Hi,

I am using `Ubuntu 20.04` with `ROS2 Foxy` and `URSim 5.11`. 

I am using the following launch command `ros2 launch ur_bringup ur_control.launch.py ur_type:=ur5e robot_ip:=127.0.0.1` to connect with URSim and able to establish connection. 

1. Is there a way to find if I am using simulation(gazebo/URSim) or real hardware from the nodes available? depending on the answer pl lmk the solution. (Also not able to find any documentation for understanding the dashboard msgs)

2.  Earlier there was `/ur_hardware_interface` node available to know this is ROS1, is it available for ROS2, how to use it?
Note: I only see `/io_and_status_controller` and able to echo the output the `robot_mode` and `robot_program_running`. But the ros2 launched terminal does not show the status called Running or Stopped that was earlier available, 

3. I am not able to see any getting started example for Gazebo, can you pl help me

I am stuck here for past few days and not able to proceed, pl. help me

Thank you

**Comment by fmauch** [comment]
Whether you use URSim or a real robot is not really distinguishable since URSim behaves exactly like a real robot. Concerning gazebo, you can have a look at https://github.com/UniversalRobots/Universal_Robots_ROS2_Gazebo_Simulation but that package is not finished / fully maintained at the moment.

**Reply by jsubramo** [reply]
Thanks for the reply! 
my concern was more reg. not able to see the ur_hardware_interface topics being published, 
can you give steps how do i fix this, 

using `ros2 launch ur_bringup ur_control.launch.py ur_type:=ur5e robot_ip:=127.0.0.1`

thank you

**Reply by fmauch** [reply]
I'm not sure I do understand your question. When you run `ros2 launch ur_robot_driver ur_control.launch.py ur_type:=ur5e robot_ip:=127.0.0.1` (You shouldn't be using `ur_bringup` anymore, as it will be gone in all distributions after humble. AND please don't use foxy anymore, it is not maintained anymore), you can see all topics using a regular `ros2 topic list` command (Note: I started with `launch_rviz:=false` in order to reduce the amount of topics):

```
/dynamic_joint_states
/force_torque_sensor_broadcaster/transition_event
/force_torque_sensor_broadcaster/wrench
/forward_position_controller/commands
/forward_position_controller/transition_event
/io_and_status_controller/io_states
/io_and_status_controller/robot_mode
/io_and_status_controller/robot_program_running
/io_and_status_controller/safety_mode
/io_and_status_controller/tool_data
/io_and_status_controller/transition_event
/joint_state_broadcaster/transition_event
/joint_states
/parameter_events
/robot_description
/rosout
/scaled_joint_trajectory_controller/controller_state
/scaled_joint_trajectory_controller/joint_trajectory
/scaled_joint_trajectory_controller/state
/scaled_joint_trajectory_controller/transition_event
/speed_scaling_state_broadcaster/speed_scaling
/speed_scaling_state_broadcaster/transition_event
/tf
/tf_static
```

If I under you are wondering why there aren't any topics published from the `ur_hardware_interface` node. First of all, no such node exists, if you have a look at the nodes being present:

```
/controller_manager
/controller_stopper
/dashboard_client
/force_torque_sensor_broadcaster
/forward_position_controller
/io_and_status_controller
/joint_state_broadcaster
/robot_state_publisher
/scaled_joint_trajectory_controller
/speed_scaling_state_broadcaster
```

All interaction on a topic level is being done by the respective controllers directly, which show up as separate nodes.

**Reply by jsubramo** [reply]
Thanks for the explanation,
I am aware of the depreciation, reason to still use foxy is I was not able to use the docker with UR Sim, facing difficulty in using URcaps in the docker, tried following the documentation, but seems very complicated for beginner.

I am developing an application and need to interact with the dashboard client to send and receive commands.

I am looking to use the following services that were available in ROS1. 

example - 
/ur_hardware_interface/dashboard/load_installation
/ur_hardware_interface/dashboard/load_program
/ur_hardware_interface/dashboard/pause
/ur_hardware_interface/dashboard/play
/ur_hardware_interface/dashboard/popup
/ur_hardware_interface/dashboard/power_off
/ur_hardware_interface/dashboard/power_on

can you pl. let know if these are available / deprecated or a work around to access them

thank you,

**Reply by fmauch** [reply]
First of all: Please have a look at our [documentation](https://docs.ros.org/en/rolling/p/ur_robot_driver/ROS_INTERFACE.html#dashboard-client).

Second: If you are looking for services, how about a `ros2 service list`?

```
$ ros2 service list
[...]
/dashboard_client/add_to_log
/dashboard_client/brake_release
/dashboard_client/clear_operational_mode
/dashboard_client/close_popup
/dashboard_client/close_safety_popup
/dashboard_client/connect
/dashboard_client/describe_parameters
/dashboard_client/get_loaded_program
/dashboard_client/get_parameter_types
/dashboard_client/get_parameters
/dashboard_client/get_robot_mode
/dashboard_client/get_safety_mode
/dashboard_client/list_parameters
/dashboard_client/load_installation
/dashboard_client/load_program
/dashboard_client/pause
/dashboard_client/play
/dashboard_client/popup
/dashboard_client/power_off
/dashboard_client/power_on
/dashboard_client/program_running
/dashboard_client/program_saved
/dashboard_client/program_state
/dashboard_client/quit
/dashboard_client/raw_request
/dashboard_client/restart_safety
/dashboard_client/set_parameters
/dashboard_client/set_parameters_atomically
/dashboard_client/shutdown
/dashboard_client/stop
/dashboard_client/unlock_protective_stop
[...]
```

**Reply by jsubramo** [reply]
Hi, sry for the delayed response, thankyou for the heads up!
(I was not able to find this doc for Foxy, so spent some time there. I will upgrade to Humble)
This is exactly what I am looking for, can you pl provide the detailed steps with launch commands to get this information, 
That would complete my query,

**Comment by jsubramo** [comment]
I was able to get it working with ROS2 Humble in Ubuntu 22. I used the UR Sim from provided docker, 
Thanks for the help.
